src/Symfony/Component/Mailer/Bridge/Amazon/Http/SesTransport.php

```diff
@@ -0,0 +1,74 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Mailer\Bridge\Amazon\Http;
+
+use Psr\Log\LoggerInterface;
+use Symfony\Component\EventDispatcher\EventDispatcherInterface;
+use Symfony\Component\HttpClient\HttpClient;
+use Symfony\Component\Mailer\Exception\TransportException;
+use Symfony\Component\Mailer\SentMessage;
+use Symfony\Component\Mailer\Transport\AbstractTransport;
+use Symfony\Contracts\HttpClient\HttpClientInterface;
+
+/**
+ * @experimental in 4.3
+ */
+class SesTransport extends AbstractTransport
+{
+    private $endpoint = 'https://email.%region%.amazonaws.com';
+    private $client;
+    private $accessKey;
+    private $secretKey;
+    private $region;
+
+    /**
+     * @param string $region Amazon SES region (currently one of us-east-1, us-west-2, or eu-west-1)
+     */
+    public function __construct(string $accessKey, string $secretKey, string $region = null, HttpClientInterface $client = null, EventDispatcherInterface $dispatcher = null, LoggerInterface $logger = null)
+    {
+        $this->client = $client ?? HttpClient::create();
+        $this->accessKey = $accessKey;
+        $this->secretKey = $secretKey;
+        $this->region = $region ?: 'eu-west-1';
+
+        parent::__construct($dispatcher, $logger);
+    }
+
+    protected function doSend(SentMessage $message): void
+    {
+        $date = gmdate('D, d M Y H:i:s e');
+        $auth = sprintf('AWS3-HTTPS AWSAccessKeyId=%s,Algorithm=HmacSHA256,Signature=%s', $this->accessKey, $this->getSignature($date));
+
+        $endpoint = \str_replace('%region%', $this->region, $this->endpoint);
+        $response = $this->client->request('POST', $endpoint, [
+            'headers' => [
+                'X-Amzn-Authorization' => $auth,
+                'Date' => $date,
+            ],
+            'body' => [
+                'Action' => 'SendRawEmail',
+                'RawMessage.Data' => \base64_encode($message->toString()),
+            ],
+        ]);
+
+        if (200 !== $response->getStatusCode()) {
+            $error = new \SimpleXMLElement($response->getContent(false));
+
+            throw new TransportException(sprintf('Unable to send an email: %s (code %s).', $error->Error->Message, $error->Error->Code));
+        }
+    }
+
+    private function getSignature($string)
+    {
+        return base64_encode(hash_hmac('sha256', $string, $this->secretKey, true));
```

Comment:
It probably would be `\base64_encode`

Reply:
no. We fully-qualify functions calls only for functions which are optimized by the compiler when it recognizes it. And that's not the case here.

Reply:
Well it is done a few lines below in the same file 😛